README
Pandamonkium Productions present
                ____        __  ___
               / __ \__  __/  |/  /___  _____
              / /_/ / / / / /|_/ / __ \/ ___/
             / ____/ /_/ / /  / / /_/ (__  )
            /_/    \__, /_/  /_/\____/____/
                  /____/ 

                                                       -- ideamonk and yuvipanda
                                                                  #hackers-india

Purpose -

    PyMos is a Python module that helps you generate a mosaic poster of an image
    using a desired collection of thumbnails.

    Some sample images can be seen here

        * http://www.flickr.com/photos/ideamonk/4020798966/sizes/m/
        * http://www.flickr.com/photos/ideamonk/4026969049/sizes/l/
        * http://www.flickr.com/photos/ideamonk/4064596012/

    Dependencies -

        * argparse          met by setuptools
        * PIL 1.1.6+        manual
        * rtree             optional
        * libspatialindex   needed by rtree module

Usage -

    1. As a command-line tool

        pymos [-h] [-z ZOOM] [-ts THUMBSIZE] [-f FUZZFACTOR] [-v] [-nc]
                    input output collection

        Creates mosaics of a collection of images to closely match a target image

        positional arguments:
        input                 Input file
        output                Output file
        collection            Directory holding images to be mosaiced

        optional arguments:
        -h, --help            show this help message and exit
        -z ZOOM, --zoom ZOOM  Zoom Level (20)
        -ts THUMBSIZE, --thumbsize THUMBSIZE
                                Size of the thumbnails (60)
        -f FUZZFACTOR, --fuzzfactor FUZZFACTOR
                                Amount of randomness in ouput (0)
        -v, --verbose         Show verbose output
        -nc, --new-colormap   Regenerate color map


    Notes:

        * 	Input files can be any standard image file supported by
            PIL (Python Imaging Library) 1.1.6+

        * 	a collection is just any folder that has many small images to build up
            the mosaic

        *	  ZOOM is the number of times input file is enlarged to make the output
            file. Eg if input file is 800x600, a zoom of 10 would give you 8000x6000

        *	  THUMBSIZE is the size of small pieces that would compose the output image.

        *	  A colormap is generated and stored for a collection (directory) in order
            to save time in regeneration. If you add new images to a collection, you
            can fire pymos with -nc option to generate a new colormap which would
            include the newly added images.

        *   How to make a collection of images ?
            Well its upto you. We just saved some random thumnails from flickr.
            You could use twitter API as well to fetch display pics of your friends
            and make a mosaic using your followers :)


    2. As a library -

        Example #1 Storing to a file

        from pymos.core import build_mosaic

        build_mosaic(
                input_path="foo.jpg",
                output_path="bar.png",
                collection_path="/tmp/apples",
                zoom=20,
                thumb_size=100,
                new_colormap=False
        )

        Example #1 Getting back an image

        from pymos.core import build_mosaic

        foo = build_mosaic(
                input_path="foo.jpg",
                output_path=None,                          # notice the None here
                collection_path="/tmp/apples",
                zoom=20,
                thumb_size=100,
                new_colormap=False
        )

        foo.show()


A more detailed, friendly description - http://huntertrek.com/wp/2009/11/30/photo-mosaics-with-pymos/

Feedback -

    Did PyMos fail over something?
    Do let us know.

    Do you wish to report a big? a feedback perhaps?
    Why don't you raise an issue on project page - http://github.com/ideamonk/PyMos/issues

    Or do you wish to make changes to suit it to your needs?
    Fork it baby - http://github.com/ideamonk/PyMos

    Look into AUTHORS for mailing us.

Btw do checkout this script to grab some images from flickr for testing it http://gist.github.com/231856


HISTORY
Pandamonkium Productions present
                ____        __  ___
               / __ \__  __/  |/  /___  _____
              / /_/ / / / / /|_/ / __ \/ ___/
             / ____/ /_/ / /  / / /_/ (__  )
            /_/    \__, /_/  /_/\____/____/
                  /____/
                                                       -- ideamonk and yuvipanda
                                                                  #hackers-india

Hi,

I was bored so I had to write something quick...

I call it PyMos, I should've called it PyMosaic, but that sounds too long.

It does something like this to input.jpg using collection/* as seen here -
http://www.flickr.com/photos/ideamonk/4020798966/sizes/l/

How to make your own -
  grab some image, put it as input.jpg besides script
  grab some matching color images, push them into collection (any folder you like)
  , keep small images
  how to - goto flickr.com search something, n.times do{ rightclick, save }
  run the script, let it work
  stare/zoom around output.png, show it to friends, feel k3wl

Then one night came YuviPanda and he hacked a little, hacked some more, and
hacked till sunrise to add most of the + on the todos

...
..
.


pymos.py
#!/usr/bin/env python

# Pandamonkium Productions present
#                ____        __  ___          
#               / __ \__  __/  |/  /___  _____
#              / /_/ / / / / /|_/ / __ \/ ___/
#             / ____/ /_/ / /  / / /_/ (__  ) 
#            /_/    \__, /_/  /_/\____/____/  
#                  /____/ 1.0
#      
#                                                      -- ideamonk and yuvipanda
#                                                                 #hackers-india

# Mosaic Generator using Python ( Diwali Night Hack )

import Image, ImageFilter
import os, sys, math, random
import logging
import glob

try:
	import cPickle as pickle
except ImportError:
	import pickle
	
def build_colormap(files):
	colormap = []
	file_count = 0
	total_files = len(files)
	log = logging.getLogger("PyMos")
	for eachfile in files:
		im = Image.open(eachfile)
		
		# lets blur a little to bring major color's prominance	
		im = im.filter(ImageFilter.BLUR)
		im = im.filter(ImageFilter.BLUR)
		im = im.filter(ImageFilter.BLUR)
	
		r = g = b = 0
		imdata = list(im.getdata())
		imdata_size = len(imdata)
		
		try:
			for i in imdata:
				r += i[0]
				g += i[1]
				b += i[2]

			# average the color of this thumbnail
			r /= imdata_size
			g /= imdata_size
			b /= imdata_size
		except:
			''' '''

		# append to colormap
		colormap.append( ( (r,g,b), eachfile, None ))
		# ^^ new format for colormap, None replaced with resized images
		# when processing to optimize
		
		file_count+=1
		log.debug("%.1f %% done" % ( (float(file_count) / total_files) * 100 ))
	return colormap


def build_mosaic(input_path, output_path, collection_path, zoom=20, thumb_size=60, new_colormap=False):
	log = logging.getLogger("PyMos")

	# Build Color Index
	log.info( "Building index...")
	
	files = glob.glob(os.path.join(collection_path, '*.jpg'))
	total_files = len(files)
	file_count = 0
	colormap_file = os.path.join(collection_path, '.colormap')
	if os.path.exists(colormap_file) and not new_colormap:
		colormap = pickle.load(open(colormap_file))
	else:
		colormap = build_colormap(files)
		pickle.dump(colormap, open(colormap_file, 'w'))
	
	log.info("Color Index built")
	
	# prepare images
	sourceImage = Image.open (input_path)
	sourceData = list(sourceImage.getdata())
	source_width, source_height = sourceImage.size
	output_width = source_width*zoom
	output_height = source_height*zoom
	
	output = Image.new(sourceImage.mode,
			   (output_width,output_height),
			   (255,255,255)
			   )

	log.info("Generating Mosaic...")
	# square mosaics as for now
	for x in xrange(0, output_width, thumb_size):
		for y in xrange(0, output_height, thumb_size):
			source_color = sourceData[ (y/zoom) * source_width + x/zoom ]
			
			# euclidean distance, color, index in colormap
			match = (555, (555,555,555), 0)# initially something out of range
		
			for index,thumbs in zip (xrange(len(colormap)), colormap):
				thumb_color, thumb_file = thumbs[0], thumbs[1]
				# calculate the euclidian distance between the two colors
				r1,g1,b1 = source_color
				r2,g2,b2 = thumb_color
				r3,g3,b3 = match[1]
			
				ecd_match = match[0]
				ecd_found = math.sqrt ( (r2-r1)**2 + (g2-g1)**2 + (b2-b1)**2)
				
				if (ecd_found < ecd_match):
						match = (ecd_found,thumb_color,index)
				
			try:
				if (colormap[match[2]][2] == None):   # has not been resized yet
					colormap[match[2]] = (colormap[match[2]][0], colormap[match[2]][1],Image.open(colormap[match[2]][1]))
					colormap[match[2]] = (colormap[match[2]][0], colormap[match[2]][1],colormap[match[2]][2].resize ((thumb_size,thumb_size),Image.BICUBIC) )
					
				output.paste (colormap[match[2]][2],(x,y))	
			except:
				''' maybe nothing got matched! '''
				
		log.debug("%.1f %% done" % ((float(x) / output_width)*100))

	log.info("Mosaic Generated")		
	output.save(output_path, "PNG")
	log.info("Output File Written")


tools/flickfetch.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Download with wget and xargs
# python flickfetch.py portrait 1 | xargs wget --directory-prefix=pics
# first argument is tag name, second is page number. 500 pics per page.
#                                                        -- Yuvi Panda

import sys
import simplejson as json
from urllib2 import urlopen
from urllib import urlencode

API_KEY = '9cd766b1c8a0e3196a9c2d2f7cb4cb01'

request_info = {'method':'flickr.photos.search',
                'license':'4,2,1,5,7',
                'sort':'interestingness-desc',
                'extras':'url_s,owner_name',
                'per_page':'500',
                'format':'json',
                'nojsoncallback':'1',
                }

request_info['tag'] = sys.argv[1]
request_info['page'] = sys.argv[2]
request_info['api_key'] = API_KEY

url = 'http://api.flickr.com/services/rest/?' + urlencode(request_info)

response = urlopen(url).read()

data = json.loads(response)

photos = data['photos']['photo']

for p in photos:
    print p['url_s']


pymos/core.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Pandamonkium Productions present
#                ____        __  ___
#               / __ \__  __/  |/  /___  _____
#              / /_/ / / / / /|_/ / __ \/ ___/
#             / ____/ /_/ / /  / / /_/ (__  )
#            /_/    \__, /_/  /_/\____/____/
#                  /____/
#
#                                                      -- ideamonk and yuvipanda
#                                                                 #hackers-india

''' PyMos - A mosaic generator module '''

import Image
import os, sys, random
import logging
import glob

try:
    import cPickle as pickle
except ImportError:
    import pickle

USING_RTREE = False
NEAREST_CACHE = {}

try:
    import rtree
    USING_RTREE = True
except:
    # we shall fall back to linear time queries in absence of rtree
    pass


def get_average_rgb(image, filename):
    red = green = blue = 0
    try:
        imdata = list(image.getdata())
        imdata_size = len(imdata)
    except:
        log.debug ("Error processing " + filename)
        return None

    try:
        for i in imdata:
            if type(i) == int:
                red += i
                green += i
                blue += i
            else:
                red += i[0]
                green += i[1]
                blue += i[2]

        # average the color of this thumbnail
        red /= imdata_size
        green /= imdata_size
        blue /= imdata_size
    except ValueError:
        log.debug ("Error processing " + filename)
    return (red, green, blue)


def get_linear_colormap(files):
    colormap = []
    file_count = 0
    total_files = len(files)
    log = logging.getLogger("PyMos")
    for eachfile in files:
        try:
            temp = Image.open(eachfile)
        except IOError as error:
            log.debug("Error opening %s" % eachfile)
            log.debug("IOError - %s" % error)
            continue

        rgb = get_average_rgb(temp, eachfile)
        if not rgb:
            continue
        
        colormap.append( ( rgb, eachfile ))
        # caching of resized images done in image_cache now

        file_count += 1
        log.debug("%.1f %% done" % ((float(file_count)/total_files)*100))
    return colormap


def get_rtree_properties():
    p = rtree.index.Property()
    p.dimension = 3
    p.dat_extension = 'data'
    p.idx_extension = 'index'
    return p
    
    
def get_rtree_colormap(files, colormap_file):
    log = logging.getLogger("PyMos")
    
    if os.path.isfile(colormap_file + ".index"):
        os.remove(colormap_file + ".index")
    if os.path.isfile(colormap_file + ".data"):
        os.remove(colormap_file + ".data")
        
    colormap = rtree.index.Index(colormap_file, properties=get_rtree_properties())
    # TODO: confirm if this is in-memory and has nothing to do with the 
    # .colormap.data and .colormap.index files and can be painlessly pickled
    
    file_count = 0
    total_files = len(files)
    
    for eachfile in files:
        try:
            temp = Image.open(eachfile)
        except IOError as error:
            log.debug("Error opening %s" % eachfile)
            log.debug("IOError - %s" % error)
            continue

        rgb = get_average_rgb(temp, eachfile)
        if not rgb:
            continue
            
        r,g,b = rgb
        # append to colormap
        colormap.add( file_count, (r,g,b,r,g,b), eachfile)
        # ^^ new format for colormap, None replaced with resized images
        # when processing to cache already loaded images

        file_count += 1
        log.debug("%.1f %% done" % ((float(file_count)/total_files)*100))
    return colormap
    
    
def build_colormap(files, colormap_file):
    ''' builds out and returns an average color to file location mapping '''
    if not USING_RTREE:
        return get_linear_colormap(files)
    else:
        return get_rtree_colormap(files, colormap_file)


def get_euclidean_match(source_color, colormap):
    # euclidean distance, color, index in colormap
    r_1, g_1, b_1 = source_color
    match = (196608, (555, 555, 555), 0) # initially something out of range
    for index, thumbs in zip (xrange(len(colormap)), colormap):
        thumb_color = thumbs[0]
        # calculate the euclidian distance between the two colors
        r_2, g_2, b_2 = thumb_color

        ecd_match = match[0]
        ecd_found = ( (r_2 - r_1) ** 2 + (g_2 - g_1) ** 2 +
                        (b_2-b_1) ** 2 )

        if (ecd_found < ecd_match):
            match = (ecd_found, thumb_color, index)
    return colormap[match[2]][1]

def build_mosaic(input_path, output_path, collection_path,
                    zoom = 20, thumb_size = 60, fuzz = 0, new_colormap=False):
    ''' Builds up the ouput using given parameters using colormap
        using input_path, output_path, collection_path, zoom=20,
              thumb_size=60, fuzz=0, new_colormap=False
    '''
    log = logging.getLogger("PyMos")
   
    # Build Color Index
    log.info( "Building index...")

    files = glob.glob(os.path.join(collection_path, '*.jpg'))
    colormap_file = os.path.join(collection_path, '.colormap')
    NEAREST_CACHE = {}
    
    if os.path.exists(colormap_file) and not new_colormap:
        if not USING_RTREE:
            colormap = pickle.load(open(colormap_file))
        else:
            colormap = rtree.index.Index(colormap_file, properties=get_rtree_properties())
    else:
        colormap = build_colormap(files, colormap_file)
        if not USING_RTREE:
            pickle.dump(colormap, open(colormap_file, 'w'))

    log.info("Color Index built")

    # prepare images
    try:
        source = Image.open (input_path)
        source_data = list(source.getdata())
    except IOError:
        log.debug ("Error opening %s" % input_path)
        sys.exit(0)

    source_width, source_height = source.size
    output_width, output_height = source_width*zoom, source_height*zoom

    output = Image.new("RGB", (output_width, output_height),
                            (255,255,255))
    image_cache = {}
    log.info("Generating Mosaic...")
    
    # square mosaics as for now
    for s_x in xrange(0, output_width, thumb_size):
        for s_y in xrange(0, output_height, thumb_size):
            source_color = source_data[ (s_y/zoom) * source_width + s_x/zoom ]
            is_bw = type(source_color) == int
            
            # we randomize source color for added fuziness
            if (fuzz!=0):
                if is_bw:
                    source_color = random.randint(-fuzz, fuzz) + source_color
                else:
                    source_color = tuple(s_x + random.randint(-fuzz, fuzz)
                                        for s_x in source_color)
                                        
            if is_bw:
                source_color = (source_color, source_color, source_color)
                
            if not USING_RTREE:
                match = get_euclidean_match(source_color, colormap)
            else:
                if source_color in NEAREST_CACHE:
                    match = NEAREST_CACHE[source_color]
                else:
                    match = NEAREST_CACHE[source_color] = \
                            colormap.nearest(source_color, 1, objects=True).next().object
            
            if not (match in image_cache):
                image_cache[match] = Image.open(match)
                ### new maxfill method
                tsize = image_cache[match].size
                # taller image -> fille width to complete square
                tsize =  (thumb_size,
                        int( round((float(tsize[1])/tsize[0]) * thumb_size)))

                if ( tsize[0] > tsize[1]):
                    # wider image -> fill height of thumb_sizexthumb_size square
                    tsize =  (int( round((float(tsize[0])/tsize[1]) * thumb_size )),
                        thumb_size)
                image_cache[match] = image_cache[match].resize(tsize)
                
            output.paste (image_cache[match], (s_x, s_y))
        log.debug("%.1f %% done" % ((float(s_x)/output_width) * 100))


    log.info("Mosaic Generated. Saving...")

    if (output_path == None):
        return output

    output.save(output_path, "PNG")
    log.info("Done " + output_path)
